01_Data/src/layers/sequelize/model/Tariff/TariffElement.ts

```diff
@@ -0,0 +1,175 @@
+    notEmpty,
+    safelyMap,
+    TariffId,
+    Time
+} from "@citrineos/base";
+import {Tariff} from "./Tariffs";
+import {ReservationRestrictionType} from "./ReservationRestrictionType";
+import {TimeColumn} from "../../types/TimeColumn";
+import {DateOnlyColumn} from "../../types/DateOnlyColumn";
+
+@Table
+export class TariffElement extends Model implements TariffElementData {
+
+  static readonly MODEL_NAME: string = Namespace.TariffElement;
+
+  public static newInstance(data: TariffElementData): TariffElement {
+    return TariffElement.build({...data});
+  }
+
+  @BelongsTo(() => Tariff)
+  declare tariff: Tariff;
+
+  @ForeignKey(() => Tariff)
+  @Column({
+    type: DataType.STRING,
+    validate: {
+      isTariffId: (value: string) => assert(isTariffId(value), `Invalid tariffId: ${value}`),
+    }
+  })
+  declare tariffId: TariffId;
+
+  @Column({
+    type: DataType.JSON,
+    allowNull: false,
+    validate: {
+      notNull: true,
+      atLeastOne: (value: PriceComponent[]) => assert(notEmpty(value), `priceComponents cannot be empty`),
+      hasNoDuplicates: (value: PriceComponent[]) => assert(!hasDuplicates(value, 'type'), 'priceComponents cannot contain duplicates'),
+    }
+  })
+  declare priceComponents: PriceComponent[];
+
+  @Column(DataType.VIRTUAL)
+  get restrictions(): TariffRestrictions {
+    return {
+      startTime: this.startTime?.hourMinute,
+      endTime: this.endTime?.hourMinute,
+      startDate: this.startDate?.yearMonthDay,
+      endDate: this.endDate?.yearMonthDay,
+      minKwh: this.minKwh,
+      maxKwh: this.maxKwh,
+      minCurrent: this.minCurrent,
+      maxCurrent: this.maxCurrent,
+      minPower: this.minPower,
+      maxPower: this.maxPower,
+      minDuration: this.minDuration,
+      maxDuration: this.maxDuration,
+      dayOfWeek: this.dayOfWeek,
+      reservation: this.reservation,
+    };
+  }
+
+  set restrictions(restrictions: TariffRestrictions) {
+    this.startTime = safelyMap(restrictions.startTime, Time.of);
+    this.endTime = safelyMap(restrictions.endTime, Time.of);
+    this.startDate = safelyMap(restrictions.startDate, DateOnly.of);
+    this.endDate = safelyMap(restrictions.endDate, DateOnly.of);
+    this.minKwh = restrictions.minKwh;
+    this.maxKwh = restrictions.maxKwh;
+    this.minCurrent = restrictions.minCurrent;
+    this.maxCurrent = restrictions.maxCurrent;
+    this.minPower = restrictions.minPower;
+    this.maxPower = restrictions.maxPower;
+    this.minDuration = restrictions.minDuration;
+    this.maxDuration = restrictions.maxDuration;
+    this.dayOfWeek = restrictions.dayOfWeek;
+    this.reservation = restrictions.reservation;
+  }
+
+  @TimeColumn()
+  private declare startTime?: Time;
+
+  @TimeColumn()
+  private declare endTime?: Time;
+
+  @DateOnlyColumn()
+  private declare startDate?: DateOnly;
+
+  @DateOnlyColumn()
+  private declare endDate?: DateOnly;
+
+  @Column({
+    type: DataType.DECIMAL(19, 4),
+    get(this: Tariff) {
```

Comment:
It seems Sequelize treats DECIMAL type as a string during data retrieval, hence this getter mapping. (I used it for every other DECIMAL column I added)

Reply:
Also, I was considering whether we should store and manage money in minor currency units to avoid floating-point errors with JS numbers? Instead of developing our own wrapper class, using a library like Dinero.js could be an option?
I'm also curious if you think it's worthwhile to create similar wrappers for energy (storing and calculating in Watthour instead of Kilowatthour), power, and electrical current. For now, I keep it as plain numbers.